core/src/Cuda/Kokkos_Cuda_Team.hpp

```diff
@@ -186,22 +186,23 @@ class CudaTeamMember {
   template <typename ReducerType>
   KOKKOS_INLINE_FUNCTION std::enable_if_t<is_reducer<ReducerType>::value>
   team_reduce(ReducerType const& reducer) const noexcept {
-    team_reduce(reducer, reducer.reference());
+    using value_type = typename ReducerType::value_type;
+    using wrapped_reducer_type =
+        typename Impl::FunctorAnalysis<Impl::FunctorPatternInterface::REDUCE,
+                                       TeamPolicy<Cuda>, ReducerType,
+                                       value_type>::Reducer;
+
+    team_reduce(wrapped_reducer_type(reducer), reducer.reference());
   }
 
-  template <typename ReducerType>
+  template <typename ReducerType, typename ValueType>
   KOKKOS_INLINE_FUNCTION std::enable_if_t<is_reducer<ReducerType>::value>
-  team_reduce(ReducerType const& reducer,
-              typename ReducerType::value_type& value) const noexcept {
+  team_reduce(ReducerType const& reducer, ValueType& value) const noexcept {
     (void)reducer;
     (void)value;
+
     KOKKOS_IF_ON_DEVICE(
-        (typename Impl::FunctorAnalysis<
-             Impl::FunctorPatternInterface::REDUCE, TeamPolicy<Cuda>,
-             ReducerType, typename ReducerType::value_type>::Reducer
-             wrapped_reducer(reducer);
-         cuda_intra_block_reduction(value, wrapped_reducer, blockDim.y);
-         reducer.reference() = value;))
+        (cuda_intra_block_reduction(value, reducer, blockDim.y);))
```

Comment:
Why is the reducer not wrapped anymore? This function is still user-facing, right?

Why don't you need to have
```
reducer.reference() = value;
```
anymore? Don't you need to assign in the other overload then?

Reply:
The assumption I was operating under was that the team member class and its member functions are not necessarily user-facing, because they are placed in `impl` namespace.
But, seeing that they are defined to be part of the public API in the [documentation](https://kokkos.org/kokkos-core-wiki/API/core/policies/TeamHandleConcept.html), it does make some of the changes I made in this PR invasive. I'll revert and fix those.